### PR TITLE
Implement login and logout in AWS auth manager

### PR DESCRIPTION
**Pull request by vincbeck**
This PR introduces the AWS auth manager. The target is to create an auth manager using AWS services: AWS Identity center and Amazon Verified Permissions. This PR handles the login and logout mechanism. Other features such as authorization will come in a separate PR. The auth manager wont be usable as such because of missing feature but since it is not documented anywhere, it should not be used by users. 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        return "aws_user" in session
+
+    def is_authorized_configuration(
+        self,
+        *,
+        method: ResourceMethod,
+        details: ConfigurationDetails | None = None,
+        user: BaseUser | None = None,
+    ) -> bool:
+        return self.is_logged_in()
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py:88]: For now, returns whether the user is logged-in. The actual authorization will be implemented in a separate PR. This is true for all the authorization APIs

**Comment by vincbeck** [comment]
[Providers Airflow 2.5 compatibility check](https://github.com/apache/airflow/actions/runs/6788678339/job/18455385993?pr=35488#logs) is failing because some of the features (e.g. `BaseUser`) are not available in Airflow 2.5 ... and rightfully so. But this is not an issue because this is something new (introducing the AWS auth manager). What's the solution here? @potiuk

File: docs/docker-stack/changelog.rst [hunk 2]
```diff
@@ -50,6 +50,9 @@ Airflow 2.7
 
 * 2.7.3
 
+  * Add ``libxmlsec1`` and ``libxmlsec1-dev`` libraries to dev PROD image and ``libxmlsec1`` library to runtime PROD
+    image as it is required by ``python3-saml`` library.
+
```
Comment (o-nikolas) [inline docs/docker-stack/changelog.rst]: Added to 2.7.3?? Shouldn't it be 2.8.0?

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    Leverages AWS services such as Amazon Identity Center and Amazon Verified Permissions to perform
+    authentication and authorization in Airflow.
+
+    :param appbuilder: the flask app builder
+    """
+
+    def __init__(self, appbuilder: AirflowAppBuilder) -> None:
+        super().__init__(appbuilder)
+        warnings.warn(
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: Maybe throw a `NotImplementedError`?

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            "The AWS auth manager is currently being built. It is not finalized. "
+            "It is not intended to be used yet."
+        )
+
+    def get_user_name(self) -> str:
+        user = self.get_user()
+        if not user:
+            self.log.error("Calling 'get_user_name()' but the user is not signed in.")
+            raise AirflowException("The user must be signed in.")
+        return user.get_user_name()
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: Could this be on the base class like `get_user_id` is? Looks pretty generic

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    def get_user(self) -> AwsAuthManagerUser | None:
+        return session["aws_user"] if self.is_logged_in() else None
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py:76]: Just for my own learning, where is `session` coming from?

File: airflow/providers/amazon/aws/auth_manager/views/auth.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    """
+
+    @cached_property
+    def idp_data(self) -> dict:
+        saml_metadata_url = conf.get_mandatory_value(CONF_SECTION_NAME, CONF_SAML_METADATA_URL_KEY)
+        return OneLogin_Saml2_IdPMetadataParser.parse_remote(saml_metadata_url)
+
+    @expose("/login")
+    def login(self):
+        """Start logging process."""
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/auth_manager/views/auth.py]:
```suggestion
        """Start login process."""
```

File: airflow/providers/amazon/aws/auth_manager/views/auth.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            # It provides an error reasons, if turned off, it does not
+            "debug": True,
+            "sp": {
+                "entityId": f"{base_url}/login_metadata",
+                "assertionConsumerService": {
+                    "url": f"{base_url}/login_callback",
+                    "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST",
+                },
+                "singleLogoutService": {
+                    "url": f"{base_url}/logout_callback",
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/auth_manager/views/auth.py:143]: I don't see login_callback defined above, where do we get that from?

File: airflow/providers/amazon/provider.yaml [hunk 8]
```diff
@@ -895,6 +896,20 @@ config:
...
+    description: |
+      This section only applies if you are using the AwsAuthManager. In other words, if you set
+      ``[core] auth_manager = airflow.providers.amazon.aws.auth_manager.aws_auth_manager.AwsAuthManager`` in
+      Airflow's configuration.
+    options:
+      saml_metadata_url:
+        description: |
+          SAML metadata XML file provided by AWS Identity Center.
+          This URL can be found in the AWS Identity Center console. Required.
+        version_added: 10.0.0
```
Comment (o-nikolas) [inline airflow/providers/amazon/provider.yaml]: Why v10? We are on 8.11 now

File: airflow/providers/amazon/provider.yaml [hunk 9]
```diff
@@ -895,6 +896,20 @@ config:
...
+        type: string
+        example: ~
```
Comment (o-nikolas) [inline airflow/providers/amazon/provider.yaml]: I think an example of what the url looks like would be helpful

File: airflow/www/security_manager.py [hunk 10]
```diff
@@ -355,3 +347,20 @@ def _is_authorized_category_menu(self, category: str) -> Callable:
...
+    """
+
+    def get_action(self, name: str) -> Action:
+        raise NotImplementedError()
+
+    def get_resource(self, name: str) -> Resource:
+        raise NotImplementedError()
+
+    def add_permissions_view(self, base_action_names, resource_name):
+        pass
```
Comment (o-nikolas) [inline airflow/www/security_manager.py:363]: I'm curious, why `pass` here but `NotImplementedError` above?

File: tests/providers/amazon/aws/auth_manager/views/test_auth.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+                "airflow.providers.amazon.aws.auth_manager.views.auth.OneLogin_Saml2_IdPMetadataParser"
+            ) as mock_parser:
+                mock_parser.parse_remote.return_value = SAML_METADATA_PARSED
+                return application.create_app(testing=True)
+
+    return factory()
+
+
+@pytest.mark.db_test
+class TestAwsAuthManagerAuthenticationViews:
```
Comment (o-nikolas) [inline tests/providers/amazon/aws/auth_manager/views/test_auth.py:72]: Nice tests!

File: docs/docker-stack/changelog.rst (hunk 2 above)
Comment (vincbeck) [reply docs/docker-stack/changelog.rst]: Correct :)

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py (hunk 3 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: The only reason why I did not do that is that force me to have constantly local changes to test the auth manager. But I guess it makes more sense, let me do that

Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: Or I can introduce a flag in the config, I think this is better

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py (hunk 4 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: Yeah I agree, I added a `get_name` method in `BaseUser` though. We cannot assume all auth managers will provide a property `username` to their users. Some might just use an email

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py (hunk 5 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py:76]: Flask session :) It is just an import

File: airflow/providers/amazon/aws/auth_manager/views/auth.py (hunk 7 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/views/auth.py:143]: `login_callback` is defined as route in this file (line 64). However, your comment make me thing we might not need `logout_callback`. Identity center does not support single logout. Which means, when users click on "log out", they get redirected to identity center where they can log out, but the actual log out does not happen. I talked to Identity center guys and it is intended. Anyway, I dont think the logout feature should not be very used with a workforce population of users

File: airflow/providers/amazon/provider.yaml (hunk 8 above)
Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: I did not know which version to specify since the actual release will happen in months. But I guess I can put 8.11 since this code (even though not used), should be in the next provider release

File: airflow/providers/amazon/provider.yaml (hunk 9 above)
Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: Good point

File: airflow/www/security_manager.py (hunk 10 above)
Comment (vincbeck) [reply airflow/www/security_manager.py:363]: Good question. This function (`add_permissions_view`) is called if [update_fab_perms](https://airflow.apache.org/docs/apache-airflow/1.10.9/configurations-ref.html#update-fab-perms) is `True`, which is the default value. If a user uses an auth manager different than FAB auth manager and does not set `update_fab_perms` to `False`, then they will get `NotImplementedError`. `pass` avoid that.

File: airflow/providers/amazon/provider.yaml [hunk 12]
```diff
@@ -91,6 +91,7 @@ dependencies:
   - redshift_connector>=2.0.888
   - sqlalchemy_redshift>=0.8.6
   - asgiref
+  - python3-saml
```
Comment (eladkal) [inline airflow/providers/amazon/provider.yaml]: Do we want it as dependency or in additional-extras ?

File: airflow/providers/amazon/provider.yaml [hunk 13]
```diff
@@ -895,6 +896,28 @@ config:
...
+      Airflow's configuration.
+    options:
+      use_experimental:
+        description: |
+          AWS auth manager is not ready to be used. Turn on this flag to use it anyway.
+          Do that at your own risk since the AWS auth manager is not in an usable state.
+        version_added: 8.11.0
+        type: boolean
+        example: "True"
+        default: "False"
```
Comment (eladkal) [inline airflow/providers/amazon/provider.yaml]: mmm this feels a bit odd?
The feature itself may be marked as experimental.
I guess my issue is more with the name of the config. I suggest `enable=True/False` rather than `use_experimental=True/False`

File: airflow/providers/amazon/provider.yaml [hunk 14]
```diff
@@ -895,6 +896,28 @@ config:
...
+    description: |
+      This section only applies if you are using the AwsAuthManager. In other words, if you set
+      ``[core] auth_manager = airflow.providers.amazon.aws.auth_manager.aws_auth_manager.AwsAuthManager`` in
+      Airflow's configuration.
+    options:
+      use_experimental:
+        description: |
+          AWS auth manager is not ready to be used. Turn on this flag to use it anyway.
+          Do that at your own risk since the AWS auth manager is not in an usable state.
+        version_added: 8.11.0
```
Comment (eladkal) [inline airflow/providers/amazon/provider.yaml]:
```suggestion
        version_added: 8.12.0
```

File: airflow/providers/amazon/provider.yaml [hunk 15]
```diff
@@ -895,6 +896,28 @@ config:
...
+        type: boolean
+        example: "True"
+        default: "False"
+      saml_metadata_url:
+        description: |
+          SAML metadata XML file provided by AWS Identity Center.
+          This URL can be found in the AWS Identity Center console. Required.
+        version_added: 8.11.0
```
Comment (eladkal) [inline airflow/providers/amazon/provider.yaml]:
```suggestion
        version_added: 8.12.0
```

File: airflow/providers/amazon/provider.yaml (hunk 13 above)
Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: Sounds good to me

File: airflow/providers/amazon/provider.yaml (hunk 12 above)
Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: What is the difference?

Comment (eladkal) [reply airflow/providers/amazon/provider.yaml]: to my perspective dependency is something that is crucial to 85%+ of the providers uses. Without it the provider has no meaning (for example `boto3`)
Extra is something that relatively small subset of users need thus must ask for it explicitly.

The motivation for it is that we need to remember that users do custom stuff on top of Airflow.
For example consider that I have some internal libraries that I install with Airflow (generate my own constraint file) the last thing I want is to handle conflicts with library that I don't even use just because it is installed by default.

Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: That's very valid and I definitely think `python3-saml` qualifies as additional-extras then

Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: Thanks for the learning

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py (hunk 3 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: A flag in config could be nice. But then is it a back compat issue to remove it in the future?

I think probably just dealing with the slight annoyance of commenting out that exception is the easiest thing to do.

File: airflow/providers/amazon/aws/auth_manager/views/auth.py (hunk 7 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/auth_manager/views/auth.py:143]: ugh, my bad, I meant logout_callback in my original message. I could see login_callback defined but not logout.

> they get redirected to identity center where they can log out, but the actual log out does not happen

What does that mean it does not happen?

**Review by o-nikolas — APPROVED** [review]
Builds are still going, but other than that, the changes look good to me

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py (hunk 3 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: > But then is it a back compat issue to remove it in the future?

Setting a config that does not exist raise an error? I dont think so no? If not, there is no back compat issue :)

File: airflow/providers/amazon/aws/auth_manager/views/auth.py (hunk 7 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/views/auth.py:143]: They do not get logged out. So basically that would work that way:
- User is signed in Airflow using AWS auth manager
- User clicks on "log out"
- User's session in Airflow is cleared, hence the user is logged out in Airflow (but the session still exist in Identity Center)
- User is redirected to Identity Center (AWS portal landing page as they call it) where they are not logged out but have the option to do so (through a link)

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py (hunk 3 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: This config will be to enable/disable this feature right? If people begin to make use of that to disable the feature, then it is removed and they can no longer control the feature anymore I think that might be back compat? But I do agree it's a stretch, more of a technicality. I'd be fine with it.

File: airflow/providers/amazon/provider.yaml (hunk 12 above)
Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: Tests are failing now because the module is not installed. How can I specify in tests that I need this extra to be installed?

File: airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py (hunk 3 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py]: It is disabled by default, so I guess users wont set this config to the default value. When ready, I'll remove this config flag, thus if some mysterious users still had set this config, it will be ignored

File: airflow/providers/amazon/provider.yaml (hunk 12 above)
Comment (eladkal) [reply airflow/providers/amazon/provider.yaml]: I assume testing should be done similar to how we test s3fs extra ?

https://github.com/apache/airflow/blob/69bac3fba897f9e7b0af642c97f9af0987a875de/airflow/providers/amazon/aws/fs/s3.py#L48-L56

https://github.com/apache/airflow/blob/4cc98ba6648b4a350f9093c4e22ad6cff87ac5e4/tests/providers/amazon/aws/fs/test_s3.py#L26

Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: Which means skipping these tests :/ I dont think there is another solution but this is not ideal

Comment (eladkal) [reply airflow/providers/amazon/provider.yaml]: ~How did AWS added aiobotocore testing?
it's also extra~

Maybe should follow `aiobotocore` case that is also an extra package for amazon provider

https://github.com/apache/airflow/pull/28850#discussion_r1108517079

Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml]: Thanks. Very helpful. But I dont think it is worth creating a new CI workflow just for this test. Tests will be skipped then (it is only one test file, I guess it is fine)

File: airflow/providers/amazon/provider.yaml [hunk 16]
```diff
@@ -897,6 +900,28 @@ config:
...
+  aws_auth_manager:
+    description: |
+      This section only applies if you are using the AwsAuthManager. In other words, if you set
+      ``[core] auth_manager = airflow.providers.amazon.aws.auth_manager.aws_auth_manager.AwsAuthManager`` in
+      Airflow's configuration.
+    options:
+      enable:
+        description: |
+          AWS auth manager is not ready to be used. Turn on this flag to use it anyway.
+          Do that at your own risk since the AWS auth manager is not in an usable state.
```
Comment (ferruzzi) [inline airflow/providers/amazon/provider.yaml:912]: :rofl:

File: airflow/www/security_manager.py [hunk 17]
```diff
@@ -355,3 +347,20 @@ def _is_authorized_category_menu(self, category: str) -> Callable:
...
             self._get_auth_manager_is_authorized_method(fab_resource_name=item) for item in items
         )
+
+    """
+    The following methods are specific to FAB auth manager. They still need to be "present" in the main
+    security manager class, but they do nothing.
+    """
+
+    def get_action(self, name: str) -> Action:
+        raise NotImplementedError()
```
Comment (ferruzzi) [inline airflow/www/security_manager.py]: Should there be a message with some context here and below?  The code that was moved into here used to have one.

Comment (vincbeck) [reply airflow/www/security_manager.py]: I can add one. This is something that should never happen though, that's why I did not specify any message but it makes no harm adding one

File: airflow/providers/amazon/aws/auth_manager/views/auth.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+    @expose("/logout")
+    def logout(self):
+        """Start logout process."""
+        session.clear()
+        saml_auth = self._init_saml_auth()
+
+        return redirect(saml_auth.logout())
+
+    @csrf.exempt
```
Comment (ferruzzi) [inline airflow/providers/amazon/aws/auth_manager/views/auth.py:70]: Nitpick, maybe this is obvious to someone with a better background in auth backends, but I have no clue what CSRF is without looking it up, or why this method should be exempt.  A comment might be helpful here.

File: airflow/providers/amazon/aws/auth_manager/views/auth.py (hunk 7 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/auth_manager/views/auth.py:143]: This is an AWS Service limitation, right?  Not something we are likely to be able to influence or implement?

**Review by ferruzzi — APPROVED** [review]
Left a couple nitpicks and a clarification question, but nothing blocking.   LGTM (pending the CI going green.. tests are currently in progress)

File: airflow/providers/amazon/aws/auth_manager/views/auth.py (hunk 7 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/views/auth.py:143]: Yes :) Nothing we can do on our side

File: airflow/providers/amazon/aws/auth_manager/views/auth.py (hunk 18 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/auth_manager/views/auth.py:70]: Just added a very simple comment. I only know I need to disable CSRF otherwise Identity center redirection wont work :)

File: airflow/auth/managers/base_auth_manager.py [hunk 19]
```diff
@@ -83,21 +83,29 @@ def get_api_endpoints(self) -> None | FlaskApi:
...
         return None
 
-    @abstractmethod
     def get_user_name(self) -> str:
         """Return the username associated to the user in session."""
+        user = self.get_user()
+        if not user:
+            self.log.error("Calling 'get_user_name()' but the user is not signed in.")
+            raise AirflowException("The user must be signed in.")
+        return user.get_name()
```
Comment (utkarsharma2) [inline airflow/auth/managers/base_auth_manager.py:92]: @vincbeck This looks like a breaking change since `get_name()` method was just defined in the BaseUser class. WDYT?

Comment (vincbeck) [reply airflow/auth/managers/base_auth_manager.py:92]: No because the auth manager interface is not public interface YET. It will become in the near future but work in auth manager is still in progress. It is not documented anywhere how to build an auth manager or even how to use it so I assume nobody uses it
